File: 22_专题/回文/medium/516. 最长回文子序列.ts
// 回文子串是要连续的，回文子序列可不是连续的！
/**
 * @param {string} s
 * @return {number}
 * dp[i][j]：字符串s在[i, j]范围内最长的回文子序列的长度为dp[i][j]。
 * 如果s[i]与s[j]相同，那么dp[i][j] = dp[i + 1][j - 1] + 2;
 */
const longestPalindromeSubseq = function (s: string): number {
  const len = s.length
  const dp = Array.from<number, number[]>({ length: len }, () => Array(len).fill(0))
  for (let i = 0; i < len; i++) {
    dp[i][i] = 1
  }

  // 倒序遍历
  for (let i = len - 1; i >= 0; i--) {
    for (let j = i + 1; j < len; j++) {
      if (s[i] === s[j]) {
        dp[i][j] = dp[i + 1][j - 1] + 2
      } else {
        dp[i][j] = Math.max(dp[i + 1][j], dp[i][j - 1])
      }
    }
  }

  return dp[0][len - 1]
}

console.log(longestPalindromeSubseq('bbbab'))

export default 1

File: 11_动态规划/力扣加加/双字符dp/516. 最长回文子序列.py
from functools import lru_cache


class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)

        @lru_cache(None)
        def dfs(l, r) -> int:
            if l > r:
                return 0
            if l == r:
                return 1
            if s[l] == s[r]:
                return dfs(l + 1, r - 1) + 2
            return max(dfs(l, r - 1), dfs(l + 1, r))

        return dfs(0, n - 1)


print(Solution().longestPalindromeSubseq(s="bbbab"))
